src/common/password.cpp

```diff
@@ -167,8 +169,9 @@ namespace
 
   void clear(std::string& pass) noexcept
   {
-    //! TODO Call a memory wipe function that hopefully is not optimized out
-    pass.replace(0, pass.capacity(), pass.capacity(), '\0');
+    // technically, the std::string documentation says the data should not be modified,
```

Comment:
Mind quoting the relevant bit of the standard? IIRC using `data()` this way has been "blessed" since c++11.

Reply:
I don't have the standard handy. I'm going by http://www.cplusplus.com/reference/string/string/data, which says, in part, "A program shall not alter any of the characters in this sequence". Note that the function returns a const char*, not a char*.

Reply:
21.4.7.1 c++11 [1]:

> const charT* data() const noexcept;
> 1 Returns: A pointer p such that p + i == &operator[](i) for each i in [0,size()].
> 2 Complexity: constant time.
> 3 Requires: The program shall not alter any of the values stored in the character array.

21.3.6 in c++98 [2]:

> const charT * data () const ;
> 3 Returns: If size() is nonzero, the member returns a pointer to the initial element of an array whose first size()
elements equal the corresponding elements of the string controlled by *this. If size() is zero, the member
returns a non-null pointer that is copyable and can have zero added to it.
> 4 Requires: The program shall not alter any of the values stored in the character array. Nor shall the program
treat the returned value as a valid pointer value after any subsequent call to a non- const member function of
basic_string that designates the same object as this.

TL;DR: using string::data() like this is fine.

1: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf

2: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2005/n1905.pdf

Reply:
I don't get it. Both seem to say "The program shall not alter any of the values stored in the character array."

Reply:
Wait, shit. I got the two of them confused. You're right.

Probably have to resort to `&pass[0]`.

Reply:
(was getting vector::data() and string::data() confused, btw)

Reply:
Do not cast away the constness here. `std::string` prior to C++11 was allowed to do CoW, and libstdc++ had such an implementation. I _believe_ gcc prior to 5.0 (which is still supported by monero) shipped with the older ABI, which still is likely doing CoW on those platforms. In that situation, this `memwipe` call could be modifying an aliased `std::string`; the non-const version of `operator[]` must be used to force an immediate copy if necessary. Also, the contents of a CoW `std::string` cannot be wiped with certainty without a custom allocator, which is what Bitcoin does (or at least used to do).

FYI, a mutable `std::string::data()` will be available for C++17, and apparently its omission has been oversight.

Reply:
Sounds like I'll just make a small class based on std::vector then.